enhancements/cluster-scope-secret-volumes/shared-resource-validation.md

```diff
@@ -0,0 +1,413 @@
+- Whenever possible prevent unsupported use of read-write volumes, as well as any future restrictions around the CSI Volume API or Shared Resource API
+- Provide pluggable means of validation of the associated Secrets and ConfigMaps of SharedSecrets and SharedConfigMaps.
+
+And provide validation failures on API object creation/modification for each of these whenever possible.
+
+### Non-Goals
+
+
+## Proposal
+
+See upstream doc 
+[https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook)
+for explanation of `ValidationAdmissionWebhooks`
+
+We will employ this plug point for inserting the code that performs any validations added.
+
+### User Stories
+
+As a developer using the Shared Resource CSI Driver
+I want validation for `readOnly: true` to be set on volumes for the Shared Resource CSI Driver to occur prior
+to the Pod Spec'able object getting persisted to etcd, so that pods are not stuck in "Creating" state waiting for a mount that will never succeed.
+Rather, I learn of the problem on the API Object PUT/POST.
+
+As an OpenShift operator maintainer or cluster administrator
+I want to reserve the `openshift-` prefix for SharedSecrets and SharedConfigMaps
+so that future OpenShift operators can create system-level shared resources and not have name conflicts 
+with items created by OCP users.
+
+As a developer, OpenShift operator maintainer, or cluster administrator
+I want options that verify the Secret or ConfigMap content in a SharedSecret or SharedConfigMap is "valid" for 
+any of the "well known" SharedSecrets or SharedConfigMaps that OCP could pre-populate the cluster with (entitlement
+secret for example).
+
+As a developer using the Shared Resource CSI Driver I want any validations to be provisioned and lifecycled in as much
+as possible the same fashion as the driver itself for consistency (i.e. via Cluster Storage Operator).
+
+As a developer or administrator using the Shared Resource CSI Driver I want sufficient metrics on the validations' prevention of incorrect Pods,
+SharedConfigMaps, and SharedSecrets, as a means of understanding the need for correcting misuse..
+
+As a developer or administrator using the Shared Resource CSI Driver I want necessary debug of the validation mechanism to be captured by
+OCP must-gather process.
+
+
+### API Extensions
+
+Quick note:  When discussing API, 'Shared Resource' on its own == SharedConfigMap / SharedSecret, and 'Reference' == the corresponding
+Secret or ConfigMap.
+
+A [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook)
+will be used to inspect all created/modified Shared Resources, and make sure their names do not start
```

Comment:
Inspecting object updates could be problematic. Let's say an item is _removed_ from our allowlist on upgrade. Then we could have a `SharedSecret` or `SharedConfigMap` that is valid on creation, but is invalid on update after the upgrade removes the .

I don't think we want to block object updates if it already exists.

Reply:
good point .... I'll reduce to create and take out the modify bit